neqo-common/src/codec.rs

```diff
@@ -472,13 +491,129 @@ impl From<Encoder> for Vec<u8> {
     }
 }
 
-impl Write for Encoder {
+#[expect(
+    clippy::unwrap_in_result,
+    reason = "successful writing to buffer needs to be guaranteed by caller"
+)]
+impl<B: io::Write> Write for Encoder<B> {
     fn write_str(&mut self, s: &str) -> fmt::Result {
-        self.buf.extend_from_slice(s.as_bytes());
+        self.buf
+            .write_all(s.as_bytes())
+            .expect("Buffer has enough capacity.");
         Ok(())
     }
 }
 
+#[expect(clippy::unnecessary_safety_doc, reason = "relevant for created object")]
+impl<'a> Encoder<Cursor<&'a mut [u8]>> {
+    /// # Safety
+    ///
+    /// Any mutable method on [`Encoder<Cursor<&mut [u8]>>`] assumes the
+    /// underlying buffer has enough capacity for the called operation. This
+    /// invariant needs to be upheld by the caller.
+    #[must_use]
+    pub fn new_borrowed_slice(buf: &'a mut [u8]) -> Self {
+        Encoder {
+            buf: Cursor::new(buf),
+        }
+    }
+}
+
+impl<'a> Encoder<&'a mut Vec<u8>> {
+    #[must_use]
+    pub fn new_borrowed_vec(buf: &'a mut Vec<u8>) -> Self {
+        Encoder { buf }
+    }
+}
+
+pub trait Buffer: io::Write {
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn as_slice(&self) -> &[u8];
+
+    // Functions needed for `Encoder::encode_vvec_with` and `Encoder::encode_vec_with`.
+
+    fn write_zeroes(&mut self, n: usize);
+
+    fn write_at(&mut self, pos: usize, data: u8);
+
+    fn rotate_right(&mut self, start: usize, count: usize);
+}
+
+impl Buffer for Vec<u8> {
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn write_zeroes(&mut self, n: usize) {
+        self.resize(self.len() + n, 0);
+    }
+
+    fn write_at(&mut self, pos: usize, data: u8) {
+        self[pos] = data;
+    }
+
+    fn rotate_right(&mut self, start: usize, count: usize) {
+        self[start..].rotate_right(count);
+    }
+
+    fn as_slice(&self) -> &[u8] {
+        self.as_ref()
+    }
```

Comment:
Should the order match the trait shape?

Reply:
Ah neat, rustanalyzer can do this.

https://github.com/mozilla/neqo/pull/2677/commits/49088fcda4d5b3326988bfae7e5cafa5f4315681